drivers: flash: stm32h7 write protection and block registers functionalities

niedzwiecki-dawid opened the pull request:
Add support for extended operations write protection and block option/control registers.

This PR depends on:
- #83266
- #83273
- #83261

niedzwiecki-dawid committed:
drivers: flash: stm32h7: add support for write protection

Add support for setting flash write protection per sector for STM32H7x
chips.

Signed-off-by: Dawid Niedzwiecki <dawidn@google.com>

niedzwiecki-dawid committed:
drivers: flash: stm32h7: add support for blocking registers

Add support for blocking flash control registers and option bytes for
STM32H7 chips.

Signed-off-by: Dawid Niedzwiecki <dawidn@google.com>

niedzwiecki-dawid commented:
Fails are caused by needed API changes #83273

niedzwiecki-dawid commented:
PR is ready to be reviewed

erwango commented on drivers/flash/Kconfig.stm32:
```diff
@@ -33,7 +33,7 @@ if SOC_FLASH_STM32
 config FLASH_STM32_WRITE_PROTECT
 	bool "Extended operation for flash write protection control"
-	depends on SOC_SERIES_STM32F4X
+	depends on SOC_SERIES_STM32F4X || SOC_SERIES_STM32H7X
```
If code also supports H7RS, it should also be populated here.
Otherwise let's remove H7RS related code below

niedzwiecki-dawid replied:
Done

duda-patryk commented on drivers/flash/flash_stm32h7x.c:
```diff
@@ -112,11 +170,161 @@ uint8_t flash_stm32_get_rdp_level(const struct device *dev)
+	uint64_t bank_mask = FLASH_WPSN_WRPSN_Msk >> FLASH_WPSN_WRPSN_Pos;
+	uint32_t pos = FLASH_WPSN_WRPSN_Pos;
+#endif
+	uint32_t sectors_per_bank = 32 - __builtin_clz(bank_mask);
```
`__builtin_clz()` counts leading 0-bits starting from most significant bit, but `bank_mask` is 64 bit variable. Is this a problem?

EDIT:
This probably works because `__builtin_clz()` takes `unsigned int` which is probably 32 bit. I think you should probably use `__builtin_clzll()`. It would be even better if you'd use `__builtin_popcountll()` which counts 1-bits.

niedzwiecki-dawid replied:
Agree, I think `__builtin_popcount()` + cast to `uint32_t` makes most sense, because `__builtin_popcount` is most standard and there is no risk of lack of implementation.

duda-patryk commented on drivers/flash/flash_stm32h7x.c:
```diff
@@ -102,6 +101,65 @@ static __unused int write_optb(const struct device *dev, uint32_t mask,
+#endif /* DUAL_BANK */
+#ifdef CONFIG_SOC_SERIES_STM32H7RSX
+	volatile uint32_t *cur =
+		(volatile uint32_t *)((uint32_t)regs + offsetof(FLASH_TypeDef, WRPSR));
```
We should use `uintptr_t` (or `intptr_t`) if you want to keep pointer as a variable.

niedzwiecki-dawid replied:
This definitely makes sense :-)

duda-patryk commented on drivers/flash/flash_stm32h7x.c:
```diff
@@ -102,6 +101,65 @@ static __unused int write_optb(const struct device *dev, uint32_t mask,
+{
+	FLASH_TypeDef *regs = FLASH_STM32_REGS(dev);
+	int rc;
+#ifdef DUAL_BANK
+	uint32_t number_of_banks = 2;
+#else  /* DUAL_BANK */
+	uint32_t number_of_banks = 1;
+#endif /* DUAL_BANK */
```
I see this is repeated below. Can we introduce some define for that e.g. `NUMBER_OF_BANKS`?

niedzwiecki-dawid replied:
Done.

duda-patryk commented on drivers/flash/flash_stm32h7x.c:
```diff
@@ -112,11 +170,161 @@ uint8_t flash_stm32_get_rdp_level(const struct device *dev)
+	uint64_t bank_mask = FLASH_WPSN_WRPSN_Msk >> FLASH_WPSN_WRPSN_Pos;
+	uint32_t pos = FLASH_WPSN_WRPSN_Pos;
+#endif
+	uint32_t sectors_per_bank = 32 - __builtin_clz(bank_mask);
```
Do we have a guarantee that each bank has the same number of sectors?

niedzwiecki-dawid replied:
Yes, from I saw in the HAL, there is always one mask defined for both banks.

niedzwiecki-dawid commented:
Rebased and changes after review.

duda-patryk commented on drivers/flash/flash_stm32h7x.c:
```diff
@@ -102,6 +104,60 @@ static __unused int write_optb(const struct device *dev, uint32_t mask,
+	/* Make sure previous write is completed. */
+	barrier_dsync_fence_full();
+
+	rc = flash_stm32_wait_flash_idle(dev);
```
From RM0433 Rev 8 p. 175
```
To modify user option bytes, follow the sequence below:

1. Unlock FLASH_OPTCR register as described in Section 4.5.1: FLASH configuration
protection, unless the register is already unlocked.
2. Write the desired new option byte values in the corresponding option registers
(FLASH_XXX_PRG1/2).
3. Set the option byte start change OPTSTART bit to 1 in the FLASH_OPTCR register.
4. Wait until OPT_BUSY bit is cleared.
```

I think we are not waiting for `OPT_BUSY` bit to clear

niedzwiecki-dawid replied:
You are so right. I sent #84430 that also fixes that problem for RDP and this PR can reuse added functions. So I'll wait till #84430 is merged

niedzwiecki-dawid replied:
Done.

niedzwiecki-dawid commented on drivers/flash/flash_stm32h7x.c:
```diff
@@ -112,11 +168,156 @@ uint8_t flash_stm32_get_rdp_level(const struct device *dev)
+	uint64_t bank_mask = FLASH_WPSN_WRPSN_Msk >> FLASH_WPSN_WRPSN_Pos;
+	uint32_t pos = FLASH_WPSN_WRPSN_Pos;
+#endif
+	uint32_t sectors_per_bank = __builtin_popcount((uint32_t)bank_mask);
```
Changed.

duda-patryk commented on drivers/flash/flash_stm32h7x.c:
```diff
@@ -137,6 +159,74 @@ void flash_stm32_set_rdp_level(const struct device *dev, uint8_t level)
+			commit = true;
+		}
+		ret = write_optwp(dev, changed_sectors_reg, protected_sectors_reg, i, commit);
+		if (ret != 0) {
+			return ret;
+		}
```
I think, error handling here is not that simple as you might expect. What would happen if the `write_optwp()` function fails for the second bank? What would happen if we fail to update the first bank, but the second would succeed? Even if we commit option bytes outside of the loop as I suggest in the other comment, returning error here will basically skip commit.

niedzwiecki-dawid replied:
Am I understanding correctly that you think it would be good to commit changes, even if call `write_optwp(dev, changed_sectors_reg, protected_sectors_reg, i, false);` fails for one of the banks? (not current version)

niedzwiecki-dawid replied:
You are right.

niedzwiecki-dawid committed:
drivers: flash: stm32h7: change return value of write_opt

Change the return value of the write_opt function. If returns 0 if a
change of option bytes was not needed.

It gives callers of the function an information a commit of the option
bytes is needed.

Signed-off-by: Dawid Niedzwiecki <dawidn@google.com>

niedzwiecki-dawid committed:
drivers: flash: stm32h7: change place of memory barrier

Move the memory barrier to a function that commits option bytes.

This way the barrier doesn't have to be added before every call of the
commit_optb function.

Signed-off-by: Dawid Niedzwiecki <dawidn@google.com>

niedzwiecki-dawid commented on drivers/flash/flash_stm32h7x.c:
```diff
@@ -137,6 +159,74 @@ void flash_stm32_set_rdp_level(const struct device *dev, uint8_t level)
+int flash_stm32_update_wp_sectors(const struct device *dev, uint64_t changed_sectors,
+				  uint64_t protected_sectors)
+{
+	uint64_t bank_mask = WP_MSK >> WP_POS;
```
Done.

niedzwiecki-dawid commented on drivers/flash/flash_stm32h7x.c:
```diff
@@ -137,6 +159,74 @@ void flash_stm32_set_rdp_level(const struct device *dev, uint8_t level)
+		protected_sectors_reg = (protected_sectors >> sectors_per_bank * i) & bank_mask;
+		changed_sectors_reg = (changed_sectors >> sectors_per_bank * i) & bank_mask;
+
+		if (changed_sectors == 0) {
```
Done.

niedzwiecki-dawid commented on drivers/flash/flash_stm32h7x.c:
```diff
@@ -143,6 +165,84 @@ void flash_stm32_set_rdp_level(const struct device *dev, uint8_t level)
+		/* Make sure to return the first error. */
+		if (ret < 0 && ret2 != 0) {
+			ret2 = ret;
+		}
```
Done.

erwango commented on tests/drivers/flash/stm32/testcase.yaml:
```diff
@@ -54,3 +54,11 @@ tests:
+      - google_icetower
+    extra_configs:
+      - CONFIG_FLASH_STM32_BLOCK_REGISTERS=y
+    filter: dt_compat_enabled("st,stm32h7-flash-controller") and
+            dt_label_with_parent_compat_enabled("storage_partition", "fixed-partitions")
```
This filter will also match for h7rs based platforms. Is that expected ?

niedzwiecki-dawid replied:
No, since it is not supported for h7rs. Isn't platform allow enough?

erwango replied:
Right, but then maybe filter on `CONFIG_SOC_SERIES_STM32H7XX` instead of this compatible which cover both series to be consistent.

niedzwiecki-dawid replied:
Done.
